lib/parsers/acorn-extensions.js

```diff
@@ -15,154 +15,180 @@ exports.enableTolerance = function (parser) {
+  // export default ...;
+  exports.default = true;
+
+  let isAsync;
+  if (parser.type === tt._function || (isAsync = parser.isAsyncFunction())) {
+    const funcNode = parser.startNode();
+    if (isAsync) {
+      parser.next();
     }
-    node.specifiers = [];
-    node.source = null;
+    parser.next();
+    node.declaration = parser.parseFunction(funcNode, "nullableID", false, isAsync);
+  } else if (parser.type === tt._class) {
+    const classNode = parser.startNode();
+    node.declaration = parser.parseClass(classNode, "nullableID");
   } else {
-    // export { x, y as z } [from '...']
-    node.declaration = null;
-    node.specifiers = this.parseExportSpecifiers(exports);
-    this.parseExportFrom(node, exports);
+    node.declaration = parser.parseMaybeAssign();
   }
-  return this.finishNode(node, "ExportNamedDeclaration");
+  parser.semicolon();
+  return parser.finishNode(node, "ExportDefaultDeclaration");
 }
 
-// Calls the given callback with the state of the parser temporarily
-// advanced by calling this.nextToken() n times, then rolls the parser
-// back to its original state and returns whatever the callback returned.
-function withLookAhead(n, callback) {
-  const old = Object.assign(Object.create(null), this);
-  while (n-- > 0) this.nextToken();
-  try {
-    return callback(this);
-  } finally {
-    Object.assign(this, old);
+function parseExportFrom(parser, node) {
+  parser.expectContextual("from");
+  node.source = parser.type === tt.string ? parser.parseExprAtom() : null;
+  parser.semicolon();
+}
+
+function parseExportImportDefault(parser, node, exports) {
+  // export def from '...';
+  const specifier = parser.startNode();
+  specifier.exported = parser.parseIdent(true);
+
+  node.specifiers = [
+    parser.finishNode(specifier, "ExportDefaultSpecifier")
+  ];
+
+  if (parser.type === tt.comma &&
+      peekNextType(parser) === tt.star) {
+    // export def, * as ns from '...';
+    parser.next();
+    const specifier = parser.startNode();
+    parser.next();
+    parseExportNamespaceAs(parser, node, specifier, exports);
   }
+  // export def, * as ns [, { x, y as z }] from '...';
+  parseExportSpecifiersMaybe(parser, node);
+
+  parseExportFrom(parser, node);
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function peekNextType(parser) {
-  return parser.withLookAhead(1, () => parser.type);
+function parseExportImportSpecifiers(parser, node, exports) {
+  // export { x, y as z } [from '...'];
+  node.declaration = null;
+  node.specifiers = parser.parseExportSpecifiers(exports);
+
+  if (parser.isContextual("from")) {
+    parseExportFrom(parser, node, exports);
+  } else {
+    parser.semicolon();
+  }
+  return parser.finishNode(node, "ExportNamedDeclaration");
 }
 
-function isExportDefaultSpecifier() {
-  return this.type === tt.name &&
-    this.withLookAhead(1, isCommaOrFrom);
+function parseExportNamespace(parser, node, exports) {
```

Comment:
No, this is a convention that we're introducing, since Acorn doesn't extract any functions for the `node.source = ...` code.

Come to think of it, since we added `parseExportFrom`, I think we should rename that to `parseExportSource` for consistency, since "from" makes a weird noun.